File: frontend/src/stores/pricing.ts
/**
 * v1.2 "What each store owes, and what each store charges" — state for the **Prices** half of the
 * `/warehouse` desk: the chain-wide wholesale rule, one item's price board, the approvals queue
 * and the month-end statement.
 *
 * Same conventions as `stores/purchasing.ts` and `stores/distribution.ts`, deliberately:
 *
 *  - **Actions never throw.** On failure they set `error` and return `null`, so a screen can put
 *    the message in a banner without wrapping every call in try/catch.
 *  - `loading` covers reads, `busy` holds a marker while a write runs, so one row can spin
 *    without freezing the board.
 *  - Writes are gated behind the warehouse store's `allowed` getter; the server enforces the same
 *    gate regardless — **everything here is warehouse admin / head office**, including the
 *    statement for a single store.
 *
 * The three price-change endpoints live under `purchasing.*` (they have since v1.0, and v1.2 adds
 * no second mechanism) but they are driven from here, because the queue this store holds is the
 * one the Approvals screen renders and the margin figures on it are pricing's.
 */
import { defineStore } from 'pinia'
import {
  pricingApi,
  type Statement,
  type StorePrices,
  type WholesaleResult,
  type WholesaleRow,
  type WholesaleSettings
} from '@/api/pricing'
import { purchasingApi, type PriceChangeRequest } from '@/api/purchasing'
import { useWarehouseStore } from '@/stores/warehouse'

interface PricingState {
  settings: WholesaleSettings | null
  /** the open item's price board, keyed so re-opening the same item does not blank the screen */
  board: StorePrices | null
  /** the wholesale rows the Prices → Wholesale list is showing */
  wholesale: WholesaleRow[]
  statement: Statement | null
  /** the pending (or filtered) approvals queue */
  requests: PriceChangeRequest[]
  loading: boolean
  busy: string | null
  error: string | null
  notice: string | null
}

function message(e: unknown): string {
  return (e as Error)?.message || 'Something went wrong'
}

export const usePricingStore = defineStore('pricing', {
  state: (): PricingState => ({
    settings: null,
    board: null,
    wholesale: [],
    statement: null,
    requests: [],
    loading: false,
    busy: null,
    error: null,
    notice: null
  }),
  getters: {
    /** May this user see any of it? Delegates to the warehouse store — the role check lives there. */
    allowed(): boolean {
      return useWarehouseStore().allowed
    },
    markupPct: (s): number => s.settings?.markup_pct ?? 50,
    currency: (s): string => s.settings?.currency || s.board?.currency || s.statement?.currency || 'USD',
    pendingRequests: (s): PriceChangeRequest[] => s.requests.filter((r) => r.workflow_state === 'Pending Approval'),
    pendingCount(): number {
      return this.pendingRequests.length
    }
  },
  actions: {
    /** Load `me` once, then ask the warehouse store's `allowed` getter. Sets `error` when refused. */
    async ensureAllowed(): Promise<boolean> {
      const warehouse = useWarehouseStore()
      if (!warehouse.me && !warehouse.meError) await warehouse.loadMe()
      if (!warehouse.allowed) {
        this.error = 'Pricing is head office’s — warehouse admin or head office only'
        return false
      }
      return true
    },
    clearNotice() {
      this.notice = null
    },
    clearError() {
      this.error = null
    },

    // ---------------------------------------------------------------- §A the chain-wide rule
    async loadSettings(force = false) {
      if (this.settings && !force) return this.settings
      this.error = null
      try {
        this.settings = await pricingApi.wholesale_settings()
        return this.settings
      } catch (e) {
        this.error = message(e)
        return null
      }
    },
    /** Set the chain markup. **0 is legal** — ship at cost; the server refuses < 0 and > 1000. */
    async setMarkup(pct: number) {
      if (!(await this.ensureAllowed())) return null
      this.busy = 'markup'
      this.error = null
      try {
        this.settings = await pricingApi.set_wholesale_markup(pct)
        this.notice = `Wholesale markup is now ${this.settings.markup_pct}% on what Houston paid`
        // every wholesale figure on screen was derived from the old rule
        if (this.wholesale.length) await this.loadWholesale(this.wholesale.map((r) => r.item_code))
        if (this.board) await this.loadBoard(this.board.item_code, this.board.price_list)
        return this.settings
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.busy = null
      }
    },
    async loadWholesale(itemCodes: string[]): Promise<WholesaleResult | null> {
      const codes = [...new Set((itemCodes || []).map((c) => (c || '').trim()).filter(Boolean))]
      if (!codes.length) {
        this.wholesale = []
        return null
      }
      this.loading = true
      this.error = null
      try {
        const out = await pricingApi.wholesale(codes)
        this.wholesale = out.items
        return out
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.loading = false
      }
    },
    /** Type a wholesale price on one item; `null` clears it and returns the item to the rule. */
    async setWholesale(itemCode: string, rate: number | null) {
      if (!(await this.ensureAllowed())) return null
      this.busy = itemCode
      this.error = null
      try {
        const out = await pricingApi.set_wholesale(itemCode, rate)
        const i = this.wholesale.findIndex((r) => r.item_code === itemCode)
        if (i >= 0) this.wholesale[i] = out.item
        else this.wholesale.unshift(out.item)
        this.notice =
          out.item.source === 'override'
            ? `${itemCode} is priced by hand at ${out.item.wholesale}`
            : `${itemCode} is back on the ${out.markup_pct}% chain rule`
        if (this.board?.item_code === itemCode) await this.loadBoard(itemCode, this.board.price_list)
        return out
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.busy = null
      }
    },

    // ---------------------------------------------------------------- §D the price board
    async loadBoard(itemCode: string, priceList = 'Standard Selling') {
      const code = (itemCode || '').trim()
      if (!code) {
        this.error = 'Choose an item to price'
        return null
      }
      this.loading = true
      this.error = null
      try {
        this.board = await pricingApi.store_prices(code, priceList)
        return this.board
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.loading = false
      }
    },
    /**
     * Raise one price change. A **reason is required** — the server throws without one, so the
     * board collects it rather than letting a manager meet that after typing eleven prices.
     */
    async raisePriceChange(itemCode: string, boutique: string, proposedRate: number, opts: { reason: string; valid_from?: string; valid_upto?: string }) {
      this.busy = `${itemCode}|${boutique}`
      this.error = null
      try {
        return await purchasingApi.request_price_change(itemCode, boutique, proposedRate, opts.reason, opts.valid_from, opts.valid_upto)
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.busy = null
      }
    },

    // ---------------------------------------------------------------- §D the approvals queue
    async loadRequests(filters: { boutique?: string; status?: string; item_code?: string; limit?: number } = {}) {
      this.loading = true
      this.error = null
      try {
        const out = await purchasingApi.price_change_requests(filters.boutique, filters.status ?? 'Pending Approval', filters.item_code, filters.limit ?? 100)
        this.requests = out.requests
        return out
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.loading = false
      }
    },
    /**
     * Approve or reject. **Approving is what creates the store-scoped pricing rule** — that is
     * v0.1 behaviour and nothing here reimplements it; the board is simply re-read afterwards so
     * it shows the new price rather than the one it was approved away from.
     */
    async decide(name: string, action: 'Approve' | 'Reject' = 'Approve', reason?: string) {
      if (!(await this.ensureAllowed())) return null
      this.busy = name
      this.error = null
      try {
        const out = await purchasingApi.approve_price_change(name, action, reason)
        const i = this.requests.findIndex((r) => r.name === name)
        const row = i >= 0 ? this.requests[i] : null
        if (i >= 0 && row) this.requests[i] = { ...row, workflow_state: out.workflow_state, pricing_rule: out.pricing_rule ?? null }
        this.notice =
          out.workflow_state === 'Approved'
            ? `${name} approved — ${row ? `${row.boutique} now sells ${row.item_code} at the new price` : 'the store pricing rule is live'}`
            : `${name} rejected — the store keeps its current price`
        if (this.board && row && this.board.item_code === row.item_code) await this.loadBoard(this.board.item_code, this.board.price_list)
        return out
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.busy = null
      }
    },

    // ---------------------------------------------------------------- §C the statement
    async loadStatement(from: string, to: string, boutique?: string | null) {
      this.loading = true
      this.error = null
      try {
        this.statement = await pricingApi.statement(from, to, boutique || null)
        return this.statement
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.loading = false
      }
    }
  }
})

File: maison_pos/api/pricing.py
"""Pricing API (v1.2) — ``maison_pos.api.pricing.*``.

Two prices, one module:

===========================  =========================================================
§A the wholesale price       ``wholesale`` ``set_wholesale`` ``wholesale_settings``
                             ``set_wholesale_markup`` — what a **store** pays Houston
§C the month-end statement   ``statement`` — what each store owes for a period
§D retail, per store         ``store_prices`` — every store's current shelf price for an
                             item, where it comes from, and the margin it makes. The
                             raising and approving of a new shelf price stays where it has
                             been since v0.1: ``maison_pos.api.purchasing``.
===========================  =========================================================

**Head office and the warehouse only.** Every endpoint here is behind
``assert_purchasing_admin`` — the same gate the buying screens use — because every one of them
either shows or is derived from what AWANZ Houston paid for its stock. A store manager is refused
all of it, including the statement for their own store.

**None of this is accounting.** No invoice, no receivable, no ageing, no payment. Stock still
moves at cost; the wholesale figure rides alongside for reporting (SPEC_v1.2 client decision 6).
"""

from __future__ import annotations

import json
from typing import Any, Optional

import frappe
from frappe import _
from frappe.utils import cint, flt, getdate, nowdate

from maison_pos.pricing import wholesale as wholesale_lib
from maison_pos.purchasing import assert_item, main_warehouse
from maison_pos.reports import store_statement
from maison_pos.scoping import assert_purchasing_admin

#: the selling price list the chain default lives on
DEFAULT_PRICE_LIST = "Standard Selling"


def _loads(value: Any, default: Any) -> Any:
	if value in (None, ""):
		return default
	if isinstance(value, str):
		try:
			return json.loads(value)
		except ValueError:
			return [v.strip() for v in value.split(",") if v.strip()] if default == [] else default
	return value


# ===========================================================================
# §A — the wholesale price
# ===========================================================================
@frappe.whitelist()
def wholesale_settings() -> dict[str, Any]:
	"""The chain-wide rule and what it is applied to.

	``{markup_pct, default_markup_pct, warehouse, currency, internal, notice}``
	"""
	assert_purchasing_admin()
	return {
		"markup_pct": wholesale_lib.markup_pct(),
		"default_markup_pct": wholesale_lib.DEFAULT_MARKUP_PCT,
		"warehouse": main_warehouse(),
		"currency": store_statement._currency(),
		"cost_basis": "Moving average valuation at the main warehouse",
		# one price for every store — per-partner terms are a multi-company concern (decision 1)
		"per_store_terms": False,
		"internal": True,
		"notice": store_statement.INTERNAL_NOTICE,
	}


@frappe.whitelist()
def set_wholesale_markup(pct: Any) -> dict[str, Any]:
	"""Set the chain-wide markup (percent of cost). Returns :func:`wholesale_settings`."""
	assert_purchasing_admin()
	wholesale_lib.set_markup_pct(pct)
	return wholesale_settings()


@frappe.whitelist()
def wholesale(item_codes: Any) -> dict[str, Any]:
	"""Resolve the wholesale price of many items at once — the price board lists 160 at a time.

	``item_codes`` may be a list or a JSON / comma-separated string. Returns
	``{markup_pct, currency, items: [{item_code, item_name, cost, override, wholesale, source,
	markup_pct, margin, margin_pct}], count}`` in the order asked for.
	"""
	assert_purchasing_admin()
	codes = [str(c).strip() for c in (_loads(item_codes, []) or []) if str(c).strip()]
	resolved = wholesale_lib.wholesale_for(codes)
	names = (
		{r.name: r.item_name for r in frappe.get_all("Item", filters={"name": ("in", codes)}, fields=["name", "item_name"], limit=len(codes) + 10)}
		if codes
		else {}
	)
	items = [{**resolved[c], "item_name": names.get(c) or c} for c in codes if c in resolved]
	return {
		"markup_pct": wholesale_lib.markup_pct(),
		"currency": store_statement._currency(),
		"warehouse": main_warehouse(),
		"items": items,
		"count": len(items),
	}


@frappe.whitelist()
def set_wholesale(item_code: str, rate: Any = None) -> dict[str, Any]:
	"""Type a wholesale price on one item; ``null`` clears it and returns the item to the rule.

	Returns the resolved row for that item — the same shape ``wholesale`` returns, so the screen
	can drop it straight back into the list.
	"""
	assert_purchasing_admin()
	row = wholesale_lib.set_override(item_code, rate)
	row["item_name"] = frappe.db.get_value("Item", item_code, "item_name") or item_code
	return {"item": row, "markup_pct": wholesale_lib.markup_pct(), "currency": store_statement._currency()}


# ===========================================================================
# §C — the month-end statement
# ===========================================================================
@frappe.whitelist()
def statement(from_date: Optional[str] = None, to_date: Optional[str] = None, boutique: Optional[str] = None) -> dict[str, Any]:
	"""What each store owes for the period — **a report, not an invoice**.

	One row per store (client decision 5), netted of shortages and damage (decision 4), showing
	cost and margin because it is an internal document (decision 3). Every store appears, with
	zeros when nothing was sent. Each store row carries a ``lines`` breakdown per item that the
	screen does not render and the CSV export does.

	The payload carries ``internal`` / ``is_invoice`` / ``creates_receivable`` / ``notice``: this
	must never be mistakable for a receivable, and the screen is expected to say so in words.
	"""
	assert_purchasing_admin()
	return store_statement.build(from_date, to_date, boutique)


# ===========================================================================
# §D — retail, per store: what the price board reads
# ===========================================================================
def _chain_price(item_code: str, price_list: str = DEFAULT_PRICE_LIST) -> float:
	return flt(
		frappe.db.get_value(
			"Item Price",
			{"item_code": item_code, "price_list": price_list, "selling": 1},
			"price_list_rate",
		)
	)


def _store_overrides(item_code: str, boutiques: list[str]) -> dict[str, dict[str, Any]]:
	"""The live store-scoped Pricing Rule per boutique, if any.

	The rule is the one ``AWANZ Price Change Request.apply_pricing_rule`` writes on approval —
	titled ``AWANZ <boutique> <item_code>``. Read, never written, from here: approving a request
	is what creates it, and that behaviour is v0.1's.
	"""
	from maison_pos.awanz_pos.doctype.awanz_price_change_request.awanz_price_change_request import pricing_rule_title

	titles = {pricing_rule_title(b, item_code): b for b in boutiques}
	if not titles:
		return {}
	today = getdate(nowdate())
	out: dict[str, dict[str, Any]] = {}
	for rule in frappe.get_all(
		"Pricing Rule",
		filters=[["title", "in", list(titles)]],
		fields=["name", "title", "rate", "disable", "valid_from", "valid_upto", "warehouse"],
		limit=len(titles) + 10,
	):
		boutique = titles.get(rule.title)
		if not boutique:
			continue
		live = not cint(rule.disable)
		if live and rule.valid_from and getdate(rule.valid_from) > today:
			live = False
		if live and rule.valid_upto and getdate(rule.valid_upto) < today:
			live = False
		out[boutique] = {
			"pricing_rule": rule.name,
			"rate": flt(rule.rate),
			"live": live,
			"valid_from": str(rule.valid_from) if rule.valid_from else None,
			"valid_upto": str(rule.valid_upto) if rule.valid_upto else None,
		}
	return out


def _pending_requests(item_code: str) -> dict[str, dict[str, Any]]:
	rows = frappe.get_all(
		"AWANZ Price Change Request",
		filters={"item_code": item_code, "workflow_state": "Pending Approval", "docstatus": 1},
		fields=["name", "boutique", "current_rate", "proposed_rate", "reason", "requested_by", "valid_from", "valid_upto"],
		order_by="modified desc",
		limit=200,
	)
	out: dict[str, dict[str, Any]] = {}
	for r in rows:
		out.setdefault(
			r.boutique,
			{
				"name": r.name,
				"current_rate": flt(r.current_rate),
				"proposed_rate": flt(r.proposed_rate),
				"reason": r.reason,
				"requested_by": r.requested_by,
				"valid_from": str(r.valid_from) if r.valid_from else None,
				"valid_upto": str(r.valid_upto) if r.valid_upto else None,
			},
		)
	return out


def margin_at(rate: Any, wholesale_rate: Any) -> dict[str, Any]:
	"""What a store makes selling at *rate* having paid *wholesale_rate*. The price board's maths.

	``margin_pct`` is ``None`` when there is no price to take a percentage of — an item the chain
	has never priced is not a 0 % margin, and a board that says 0 % is the sort of thing somebody
	prices against.
	"""
	rate = flt(rate)
	cost = flt(wholesale_rate)
	margin = flt(rate - cost, 2)
	return {"margin": margin, "margin_pct": flt(100.0 * margin / rate, 1) if rate else None, "has_price": bool(rate)}


@frappe.whitelist()
def store_prices(item_code: str, price_list: str = DEFAULT_PRICE_LIST) -> dict[str, Any]:
	"""Every enabled store's current shelf price for one item — the **Prices** board (§D).

	Per store: the price in force, where it comes from (a store override or the chain default),
	what the store pays Houston for the item, and the margin that price makes. Plus any request
	already waiting for approval, so the board does not invite a second one for the same row.

	Read only. Typing a new price raises ``purchasing.request_price_change``; approving it is
	``purchasing.approve_price_change``, and approval is what creates the store's Pricing Rule.
	"""
	assert_purchasing_admin()
	assert_item(item_code)
	from maison_pos.shipping import store_boutiques

	codes = store_boutiques()
	stores = frappe.get_all(
		"AWANZ Store", filters={"name": ("in", codes or ["__none__"])}, fields=["name", "boutique_name", "warehouse"], order_by="name"
	)
	resolved = wholesale_lib.wholesale_for([item_code])[item_code]
	default_rate = _chain_price(item_code, price_list)
	overrides = _store_overrides(item_code, [s.name for s in stores])
	pending = _pending_requests(item_code)

	rows = []
	for s in stores:
		override = overrides.get(s.name)
		if override and override["live"]:
			rate, source = override["rate"], "Store override"
		else:
			rate, source = default_rate, "Chain default"
		row = {
			"boutique": s.name,
			"boutique_name": s.boutique_name or s.name,
			"warehouse": s.warehouse,
			"rate": flt(rate),
			"source": source,
			"is_override": source == "Store override",
			"pricing_rule": (override or {}).get("pricing_rule"),
			"valid_from": (override or {}).get("valid_from"),
			"valid_upto": (override or {}).get("valid_upto"),
			"wholesale": resolved["wholesale"],
			"pending": pending.get(s.name),
		}
		row.update(margin_at(rate, resolved["wholesale"]))
		rows.append(row)

	item = frappe.db.get_value("Item", item_code, ["item_name", "stock_uom", "maison_barcode", "image", "item_group"], as_dict=True) or {}
	return {
		"item_code": item_code,
		"item_name": item.get("item_name") or item_code,
		"item_group": item.get("item_group"),
		"uom": item.get("stock_uom"),
		"barcode": item.get("maison_barcode"),
		"image": item.get("image"),
		"price_list": price_list,
		"default_rate": default_rate,
		"currency": store_statement._currency(),
		# what the store pays us — the same figure for every store (client decision 1)
		"wholesale": resolved["wholesale"],
		"wholesale_source": resolved["source"],
		"cost": resolved["cost"],
		"markup_pct": resolved["markup_pct"],
		"stores": rows,
		"count": len(rows),
		"internal": True,
		"notice": _("Cost and wholesale are internal AWANZ figures — do not put them in front of a store."),
	}
